util/argo/argo.go

```diff
@@ -260,6 +260,45 @@ func enrichSpec(spec *argoappv1.ApplicationSpec, appDetails *apiclient.RepoAppDe
 	}
 }
 
+// ValidateDestination checks:
+// if we used destination name instead of server to we fill the server part if exists
+// if we used both name and server then we need to see that a matching cluster exists
+func ValidateDestination(ctx context.Context, dest *argoappv1.ApplicationDestination, db db.ArgoDB) *argoappv1.ApplicationCondition {
+	if dest.Name != "" {
+		if dest.Server == "" {
+			server, err := getDestinationServer(ctx, db, dest.Name)
+			if err != nil {
+				return &argoappv1.ApplicationCondition{
+					Type:    argoappv1.ApplicationConditionInvalidSpecError,
+					Message: fmt.Sprintf("Unable to get destination server: %v", err),
+				}
+			}
+			if server == "" {
+				return &argoappv1.ApplicationCondition{
+					Type:    argoappv1.ApplicationConditionInvalidSpecError,
+					Message: fmt.Sprintf("Application references destination cluster %s which does not exist", dest.Name),
+				}
+			}
+			dest.Server = server
```

Comment:
Great idea to infer the server on the fly! This way we get very few code changes. 

I've noticed could unwanted side effects. The controller persists every app field change after app processing, so app CRD gets `server` field immediately after first reconciliation. The same thing is happening in the API server. The `validateAndNormalizeApp` populates the `server` field and application is saved with both `name`  and `server`.

Still not sure what is the best way to fix it. One idea is to create a flag field that indicates that server was inferred and use that field in a custom json marshaler:

```
// ApplicationDestination contains deployment destination information
type ApplicationDestination struct {
	// Server overrides the environment server value in the ksonnet app.yaml
	Server string `json:"server,omitempty" protobuf:"bytes,1,opt,name=server"`
	// Namespace overrides the environment namespace value in the ksonnet app.yaml
	Namespace string `json:"namespace,omitempty" protobuf:"bytes,2,opt,name=namespace"`
	// Name of the destination cluster which can be used instead of server (url) field
	Name string `json:"name,omitempty" protobuf:"bytes,3,opt,name=name"`

	isServerInferred bool `json:"-"`
}

func (d *ApplicationDestination) SetInferredServer(server string) {
	d.isServerInferred = true
	d.Server = server
}

func (d *ApplicationDestination) MarshalJSON() ([]byte, error) {
	type Alias ApplicationDestination
	dest := d
	if d.isServerInferred {
		dest = dest.DeepCopy()
		dest.Server = ""
	}
	return json.Marshal(&struct{ *Alias }{Alias: (*Alias)(dest)})
}
```

The `ValidateDestination` method should use `SetInferredServer` to indicate that server is inferred and should not be saved.

Reply:
I wanted the `server` field to be saved, because only that way I don't need to change anything in the AppProject RBAC validation. And not only that, but there is a lot of logic that expects the `server` field to have a value, many functions get it as parameter. So I would like to calculate it and keep it.

What if we do something like this: ask always for `server` while `name` to be optional, so when both are passed we validate if they matche the same cluster, if not throw spec error, if ok then the server url is used forward (no code changes there). Don't allow only name to be passed.

This will simplify code changes, but still make it clear what cluster you are using by having the name

Reply:
I think I'm proposing almost the same behavior: `server` field before processing the application so all code that uses the `server` field could stay untouched. The only difference is that `server` field should not be persisted at the same time with `name` and be inferred on the fly instead.

If `server` and `name` is stored at the same time then we need to start watching for consistency between `name` and `server`. E.g. someone might update cluster which name is referenced. In this case `name` and `server` became inconsistent. 

I've realized that API server also depends on `server` and we would have to resolve server field there as well. So I'm suggesting to add `argo.GetApplication(name, appLister, projLister)` method. The method should load application and optionally resolve server field.. 


GetApplication should be used in each application.go API method implementation and in `processAppOperationQueueItem` and `processAppRefreshQueueItem` controller methods.

@lcostea, let me know what you think please. Happy to discuss the idea in our slack channel

Reply:
I am going to implement this `argo.GetApplication(name, appLister, projLister)` and see how it will look like. Then we can decide maybe watching for inconsistency on name/server will be easier.